docs/architecture/adr-039-epoched-staking.md

```diff
@@ -0,0 +1,110 @@
+# ADR 039: Epoched Staking
+
+## Changelog
+
+- 10-Feb-2021: Initial Draft
+
+## Authors
+
+- Dev Ojha (@valardragon)
+- Sunny Aggarwall (@sunnya97)
+
+## Status
+
+Proposed
+
+## Abstract
+
+This ADR updates the proof of stake module to buffer the staking weight updates for a number of blocks before updating the consensus' staking weights. The length of the buffer is dubbed an epoch. The prior functionality of the staking module is then a special case of the abstracted module, with the epoch being set to 1 block.
+
+## Context
+
+The current proof of stake module takes the design decision to apply staking weight changes to the consensus engine immediately. This means that delegations and unbonds get applied immediately to the validator set. This decision was primarily done as it was implementationally simplest, and because we at the time believed that this would lead to better UX for clients.
+
+An alternative design choice is to allow buffering staking updates (delegations, unbonds, validators joining) for a number of blocks. This 'epoch'd proof of stake consensus provides the guarantee that the consensus weights for validators will not change mid-epoch, except in the event of a slash condition. 
+
+The decision to have immediate execution of staking changes was primarily done as it was implementationally simplest, and because we at the time believed that this would lead to better UX for clients. The UX hurdle may not be as significant as was previously thought, since it is possible to provide users acknowledgement that their bond was recorded and will be executed.
```

Comment:
```suggestion
The UX hurdle may not be as significant as was previously thought, since it is possible to provide users acknowledgement that their bond was recorded and will be executed.
```

Reply:
same sentence as above